### PR TITLE
Add whitelisting support for regex filters

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/FTL/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**

## 10

---

This PR adds more extensive whitelisting support for *FTL*DNS's RegEx filtering in a novel way: *After* regex filters have been applied, *FTL*DNS will *again* check if a domain is set for explicit whitelisting.

This should finally implement a 100% flexible solution for blocking and un-blocking of stuff.

Example 1:
You want to block all Internet and permit only `kidsafe.com`. This wasn't possible before, however, *FTL*DNS supports this now. You just add `.*` as regex (matches everything) and put `kidsace.com` on the whitelist.

Example 2:
You want to block `adserver.com` but want to be able to access `some.handy.tool.adserver.com`. Fine, just add `adserver\.com$` as regex filter and put `some.handy.tool.adserver.com` into your whitelist - done!

A note on possible performance impact: Domains that aren't matched by a regex filter are not compared against the list of whitelisted domains, so there is no cost involved for them. When a filter match happens, the check against the whitelisted domains in only done. The result is stored in memory and will be used from this point on.
TL;DR: The performance impact is expected to be minimal.

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: regex.c [hunk 1]
```diff
@@ -107,6 +142,61 @@ void free_regex(void)
...
+	free_whitelist_domains();
+}
+
+static void read_whitelist_from_file(void)
+{
+	FILE *fp;
+	char *buffer = NULL;
+	size_t size = 0;
+
+	// Get number of lines in the regex file
```
Comment (Mcat12) [inline regex.c]: These comments still need to be adjusted (they say "regex" instead of "whitelist")

File: regex.c [hunk 2]
```diff
@@ -69,6 +96,14 @@ bool match_regex(char *input)
 		}
 	}
 
+	// If a regex filter matched, we additionally compare the domain
+	// against all known whitelisted domains to possibly prevent blocking
+	// of a specific domain. The logic herein is:
+	// If matched, then compare against whitelist
+	// If in whitelist, negate matched so this function returns: not-to-be-blocked
+	if(matched)
+		matched = !in_whitelist(input);
```
Comment (Mcat12) [inline regex.c]: Should this get moved to the dnsmasq hook code? It pollutes the purpose of the method a bit. Instead of just matching against the regex, it also handles some blocking logic.

File: dnsmasq_interface.c [hunk 3]
```diff
@@ -184,7 +184,13 @@ void FTL_new_query(unsigned int flags, char *name, struct all_addr *addr, char *
...
 		// - this domain has not already been validated against the regex.
 		// This effectively prevents multiple evaluations of the same domain
-		if(match_regex(domain))
+		//
+		// If a regex filter matched, we additionally compare the domain
+		// against all known whitelisted domains to possibly prevent blocking
+		// of a specific domain. The logic herein is:
+		// If matched, then compare against whitelist
+		// If in whitelist, negate matched so this function returns: not-to-be-blocked
+		if(match_regex(domain) && !in_whitelist(domain))
```
Comment (DL6ER) [inline dnsmasq_interface.c]: Using short-circuit evaluation to ensure `in_whitelist()` is only called when `match_regex()` was `true`.
